File: Cpp/Algorithm/src/Chapter3/SwapPairs_24.cpp
#include "chapter3.h"
ListNode* myAlgo::SwapPairs_24::swapPairs(ListNode* head) {
	ListNode* dummyHead = new ListNode(0);
	dummyHead->next = head;
	// 1->2->3->4 => 2->1->3->4 => 2->1->4->3
	// 1->2 => 2->1
	ListNode* prev = dummyHead;
	ListNode* curr = head;
	while (curr!=NULL && curr->next!=NULL)
	{
		ListNode* next =curr->next;
		ListNode* nNext = next->next;

		prev->next = next;
		next->next = curr;
		curr->next = nNext;
		prev = curr;
		curr = nNext;

	}
	return dummyHead->next;
}

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/ThirdRound/Chapter03/swapPairs_24.java
package algo.tzashinorpu.ThirdRound.Chapter03;

import algo.tzashinorpu.ThirdRound.ListNode;

public class swapPairs_24 {
    public ListNode swapPairs(ListNode head) {
        // 1->2->3->null 2->1->3->null
        // 1->2->null 2->1->null
        ListNode curr = head;
        ListNode dummyHead = new ListNode();
        dummyHead.next = head;
        ListNode prev = dummyHead;
        while (curr != null && curr.next != null) {
            ListNode next = curr.next;
            ListNode nnext = next.next;
            prev.next = next;
            next.next = curr;
            curr.next = nnext;
            prev = curr;
            curr = nnext;
        }
        return dummyHead.next;
    }
}
